sdk/helper/ocsp/client.go

```diff
@@ -0,0 +1,958 @@
+			ctx, ocspClient, http.NewRequest, u, headers, ocspReq, issuer)
+		if err != nil {
+			return nil, err
+		}
+		if ocspS.code != ocspSuccess {
+			return ocspS, nil
+		}
+
+		ret, err = validateOCSP(ocspRes)
+		if err != nil {
+			return nil, err
+		}
+		if isValidOCSPStatus(ret.code) {
+			break
+		}
+	}
+	if !isValidOCSPStatus(ret.code) {
+		return ret, nil
+	}
+	v := ocspCachedResponse{time: float64(time.Now().UTC().Unix()), resp: base64.StdEncoding.EncodeToString(ocspResBytes)}
+	c.ocspResponseCacheLock.Lock()
+	c.ocspResponseCache[*encodedCertID] = &v
+	c.cacheUpdated = true
+	c.ocspResponseCacheLock.Unlock()
+	return ret, nil
+}
+
+func isValidOCSPStatus(status ocspStatusCode) bool {
+	return status == ocspStatusGood || status == ocspStatusRevoked || status == ocspStatusUnknown
+}
+
+// VerifyPeerCertificate verifies all of certificate revocation status
+func (c *Client) VerifyPeerCertificate(ctx context.Context, verifiedChains [][]*x509.Certificate, extraCas []*x509.Certificate, ocspServersOverride []string) (err error) {
+	for i := 0; i < len(verifiedChains); i++ {
+		// Certificate signed by Root CA. This should be one before the last in the Certificate Chain
+		numberOfNoneRootCerts := len(verifiedChains[i]) - 1
+		if !verifiedChains[i][numberOfNoneRootCerts].IsCA || string(verifiedChains[i][numberOfNoneRootCerts].RawIssuer) != string(verifiedChains[i][numberOfNoneRootCerts].RawSubject) {
+			// Check if the last Non Root Cert is also a CA or is self signed.
+			// if the last certificate is not, add it to the list
+			rca := c.caRoot[string(verifiedChains[i][numberOfNoneRootCerts].RawIssuer)]
+			if rca == nil {
+				return fmt.Errorf("failed to find root CA. pkix.name: %v", verifiedChains[i][numberOfNoneRootCerts].Issuer)
+			}
+			verifiedChains[i] = append(verifiedChains[i], rca)
+			numberOfNoneRootCerts++
+		}
+		results, err := c.GetAllRevocationStatus(ctx, verifiedChains[i], extraCas, ocspServersOverride)
+		if err != nil {
+			return err
+		}
+		if r := c.canEarlyExitForOCSP(results, numberOfNoneRootCerts); r != nil {
+			return r.err
+		}
+	}
+
+	return nil
+}
+
+func (c *Client) canEarlyExitForOCSP(results []*ocspStatus, chainSize int) *ocspStatus {
+	msg := ""
+	if atomic.LoadUint32((*uint32)(&c.ocspFailOpen)) == (uint32)(OCSPFailOpenFalse) {
+		// Fail closed. any error is returned to stop connection
+		for _, r := range results {
+			if r.err != nil {
+				return r
+			}
+		}
+	} else {
+		// Fail open and all results are valid.
+		allValid := len(results) == chainSize
+		for _, r := range results {
+			if !isValidOCSPStatus(r.code) {
+				allValid = false
+				break
+			}
+		}
+		for _, r := range results {
+			if allValid && r.code == ocspStatusRevoked {
+				return r
+			}
+			if r != nil && r.code != ocspStatusGood && r.err != nil {
+				msg += "" + r.err.Error()
+			}
+		}
+	}
+	if len(msg) > 0 {
+		c.Logger().Warn(
+			"WARNING!!! Using fail-open to connect. Driver is connecting to an "+
```

Comment:
This message might need some tweaking as it will be odd reading this from a cert auth point of view?